File: crates/pray-core/src/constraint.rs
use crate::{PrayError, PrayResult};
use semver::{Version, VersionReq};

/// Normalizes a Prayfile version constraint per SPEC §16.
///
/// Bare semver strings such as `1.0.0` are exact pins (`=1.0.0`), not caret ranges.
pub fn normalize_version_constraint(constraint: &str) -> String {
    let trimmed = constraint.trim();
    if trimmed.is_empty() || trimmed == "*" {
        return trimmed.to_string();
    }
    if trimmed.starts_with("~>")
        || trimmed.starts_with('~')
        || trimmed.starts_with('^')
        || trimmed.starts_with('=')
        || trimmed.starts_with('>')
        || trimmed.starts_with('<')
        || trimmed.contains('*')
    {
        return trimmed.to_string();
    }
    if Version::parse(trimmed).is_ok() {
        return format!("={trimmed}");
    }
    trimmed.to_string()
}

pub fn version_satisfies(version: &str, constraint: &str) -> PrayResult<bool> {
    let normalized = normalize_version_constraint(constraint);
    if normalized.is_empty() || normalized == "*" {
        return Ok(true);
    }
    let version =
        Version::parse(version).map_err(|error| PrayError::Resolution(error.to_string()))?;
    let req = if normalized.trim_start().starts_with("~>") {
        VersionReq::parse(&ruby_pessimistic_to_semver(&normalized)?)
            .map_err(|error| PrayError::Resolution(error.to_string()))?
    } else {
        VersionReq::parse(normalized.trim())
            .map_err(|error| PrayError::Resolution(error.to_string()))?
    };
    Ok(req.matches(&version))
}

/// Builds a Ruby pessimistic constraint (`~>`) that allows the given release line.
pub fn pessimistic_constraint_for_version(version: &str) -> PrayResult<String> {
    let parsed =
        Version::parse(version).map_err(|error| PrayError::Resolution(error.to_string()))?;
    if parsed.minor == 0 && parsed.patch == 0 {
        Ok(format!("~> {}.0", parsed.major))
    } else {
        Ok(format!("~> {}.{}", parsed.major, parsed.minor))
    }
}

/// Derives a Prayfile constraint that admits `latest_version`, preserving operator style.
pub fn latest_constraint_for_package(
    current_constraint: &str,
    latest_version: &str,
) -> PrayResult<String> {
    let normalized = normalize_version_constraint(current_constraint);
    if normalized == "*" {
        return Ok("*".to_string());
    }
    if normalized.starts_with("~") {
        return pessimistic_constraint_for_version(latest_version);
    }
    if normalized.starts_with('^') {
        let parsed = Version::parse(latest_version)
            .map_err(|error| PrayError::Resolution(error.to_string()))?;
        return Ok(format!("^{}.{}", parsed.major, parsed.minor));
    }
    if normalized.starts_with('=') || Version::parse(current_constraint.trim()).is_ok() {
        return Ok(format!("={latest_version}"));
    }
    pessimistic_constraint_for_version(latest_version)
}

fn ruby_pessimistic_to_semver(constraint: &str) -> PrayResult<String> {
    let text = constraint.trim().trim_start_matches("~>").trim();
    let parts: Vec<&str> = text.split('.').collect();
    if parts.is_empty() || parts.len() > 3 {
        return Err(PrayError::Resolution(format!(
            "unsupported Ruby pessimistic constraint: {constraint}"
        )));
    }
    let mut numbers = [0u64; 3];
    for (index, part) in parts.iter().enumerate() {
        numbers[index] = part
            .parse::<u64>()
            .map_err(|error| PrayError::Resolution(error.to_string()))?;
    }
    let lower = format!("{}.{}.{}", numbers[0], numbers[1], numbers[2]);
    let upper = match parts.len() {
        1 => format!("{}.0.0", numbers[0] + 1),
        2 => format!("{}.{}.0", numbers[0], numbers[1] + 1),
        _ => format!("{}.{}.0", numbers[0], numbers[1] + 1),
    };
    Ok(format!(">={}, <{}", lower, upper))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bare_semver_is_exact_pin() {
        assert_eq!(normalize_version_constraint("1.0.0"), "=1.0.0");
        assert_eq!(normalize_version_constraint("  2.3.4  "), "=2.3.4");
    }

    #[test]
    fn explicit_operators_are_preserved() {
        assert_eq!(normalize_version_constraint("~> 1.0"), "~> 1.0");
        assert_eq!(normalize_version_constraint("^2.0"), "^2.0");
        assert_eq!(normalize_version_constraint("= 1.2.3"), "= 1.2.3");
        assert_eq!(normalize_version_constraint("*"), "*");
    }

    #[test]
    fn bare_semver_matches_only_exact_version() {
        assert!(version_satisfies("1.0.0", "1.0.0").expect("matches"));
        assert!(!version_satisfies("1.0.1", "1.0.0").expect("does not match"));
        assert!(version_satisfies("1.0.1", "~> 1.0").expect("pessimistic matches"));
    }

    #[test]
    fn pessimistic_constraint_uses_major_minor_line() {
        assert_eq!(
            pessimistic_constraint_for_version("2.0.1").expect("constraint"),
            "~> 2.0"
        );
        assert_eq!(
            pessimistic_constraint_for_version("1.4.3").expect("constraint"),
            "~> 1.4"
        );
        assert!(version_satisfies("2.0.0", "~> 2.0").expect("matches"));
    }

    #[test]
    fn latest_constraint_preserves_operator_family() {
        assert_eq!(
            latest_constraint_for_package("~> 1.0", "2.0.0").expect("constraint"),
            "~> 2.0"
        );
        assert_eq!(
            latest_constraint_for_package("1.0.0", "2.0.0").expect("constraint"),
            "=2.0.0"
        );
        assert_eq!(
            latest_constraint_for_package("^1.0", "2.1.0").expect("constraint"),
            "^2.1"
        );
    }
}

File: npmjs/pray-cli/src/constraint.ts
import semver, { type Range, type SemVer } from "semver";
import { PrayError } from "./errors.js";

export function normalizeVersionConstraint(constraint: string): string {
  const trimmed = constraint.trim();
  if (trimmed.length === 0 || trimmed === "*") {
    return trimmed;
  }
  if (
    trimmed.startsWith("~>") ||
    trimmed.startsWith("~") ||
    trimmed.startsWith("^") ||
    trimmed.startsWith("=") ||
    trimmed.startsWith(">") ||
    trimmed.startsWith("<") ||
    trimmed.includes("*")
  ) {
    return trimmed;
  }
  if (semver.valid(trimmed)) {
    return `=${trimmed}`;
  }
  return trimmed;
}

export function versionSatisfies(version: string, constraint: string): boolean {
  const normalized = normalizeVersionConstraint(constraint);
  if (normalized.length === 0 || normalized === "*") {
    return true;
  }
  const parsedVersion = semver.parse(version);
  if (!parsedVersion) {
    throw PrayError.resolution(`invalid version ${version}`);
  }
  const requirement = normalized.trimStart().startsWith("~>")
    ? rubyPessimisticToSemver(normalized)
    : normalized.trim();
  const range = semver.validRange(requirement);
  if (!range) {
    throw PrayError.resolution(`invalid constraint ${constraint}`);
  }
  return semver.satisfies(parsedVersion, range);
}

export function pessimisticConstraintForVersion(version: string): string {
  const parsed = semver.parse(version);
  if (!parsed) {
    throw PrayError.resolution(`invalid version ${version}`);
  }
  if (parsed.minor === 0 && parsed.patch === 0) {
    return `~> ${parsed.major}.0`;
  }
  return `~> ${parsed.major}.${parsed.minor}`;
}

function rubyPessimisticToSemver(constraint: string): string {
  const text = constraint.trim().replace(/^~>\s*/, "");
  const parts = text.split(".");
  if (parts.length === 0 || parts.length > 3) {
    throw PrayError.resolution(
      `unsupported Ruby pessimistic constraint: ${constraint}`,
    );
  }
  const numbers = [0, 0, 0];
  for (let index = 0; index < parts.length; index += 1) {
    const parsed = Number.parseInt(parts[index]!, 10);
    if (Number.isNaN(parsed)) {
      throw PrayError.resolution(`invalid constraint segment ${parts[index]}`);
    }
    numbers[index] = parsed;
  }
  const first = numbers[0];
  const second = numbers[1];
  const third = numbers[2];
  if (first === undefined || second === undefined || third === undefined) {
    throw PrayError.resolution(
      `invalid constraint segment count in ${constraint}`,
    );
  }
  const lower = `${first}.${second}.${third}`;
  const upper =
    parts.length === 1 ? `${first + 1}.0.0` : `${first}.${second + 1}.0`;
  return `>=${lower} <${upper}`;
}

export type { Range, SemVer };
